base/range.jl

```diff
@@ -536,10 +498,25 @@ function getindex{T}(r::FloatRange{T}, i::Integer)
     convert(T, (r.start + (i-1)*r.step)/r.divisor)
 end
 
-function getindex{T}(r::LinSpace{T}, i::Integer)
+function getindex(r::LinSpace, i::Integer)
     @_inline_meta
     @boundscheck checkbounds(r, i)
-    convert(T, ((r.len-i)*r.start + (i-1)*r.stop)/r.divisor)
+    unsafe_getindex(r, i)
+end
+
+# This is separate to make it useful even when running with --check-bounds=yes
+function unsafe_getindex(r::LinSpace, i::Integer)
+    d = r.lendiv
+    j, a, b = ifelse(2i >= length(r), (i-1, r.start, r.stop), (length(r)-i, r.stop, r.start))
+    lerpi(j, d, a, b)
+end
+
+# High-precision interpolation. Accurate for t ∈ [0.5,1], so that 1-t is exact.
+function lerpi{T}(j::Integer, d::Integer, a::T, b::T)
+    @_inline_meta
+    t = j/d
+    # computes (1-t)*a + t*b
+    T(fma(t, b, fma(-t, a, a)))
```

Comment:
To address the "confused": the point being this is the fallback implementation, so should support any type that implements `fma`. `T` might not be a number. Or, `T` might be a number type like `Fixed{Int8,7}` (from FixedPointNumbers) which can't represent `1`. Linear interpolation between two valid numbers is still perfectly fine in such circumstances, though, so there's no need to fail. That's why I'm not going to take your suggestion :smile:.